Source/AuthenticationInterceptor.swift

```diff
@@ -0,0 +1,324 @@
+    /// If your authentication server will never reject non-expired credentials for performance reasons, then you can
+    /// safely return `false` for this method.
+    func didRequest(_ urlRequest: URLRequest, with response: HTTPURLResponse, failDueToAuthenticationError error: Error) -> Bool
+
+    /// NOTE: CN rework docstring
+    ///
+    /// Returns whether the `URLRequest` was authenticated with the specified `Credential`. If it was not, then we know
+    /// the `URLRequest` was authenticated with a previous `Credential` and can be immediately retried with the new
+    /// `Credential`.
+    ///
+    /// This is an edge case that can occur if the authentication server can invalidate non-expired credentials. In
+    /// this case, it's possible for requests to be executed with non-expired, invalid credentials. Refresh will be
+    /// triggered when the first failing response comes back from the authentication server. It is possible that some
+    /// of the requests will not return from the authentication server until the refresh is completed. In this case,
+    /// we inspect the slower requests when returned from the authentication server to see if the credential applied
+    /// to them is the same as the latest credential. If not, then we know we've refreshed the credential while the
+    /// request was in flight and it is immediately retried with the new credential. If yes, then we know we've found
+    /// the first request to fail due to a non-expired, invalidate credential and refresh is triggered.
+    func isRequest(_ urlRequest: URLRequest, authenticatedWith credential: AuthCredential) -> Bool
+}
+
+// MARK: -
+
+/// NOTE: CN docstring
+public enum AuthenticationError: Error {
+    /// NOTE: CN docstring
+    case missingCredential
+
+    /// NOTE: CN docstring
+    case excessiveRefresh
+}
+
+// MARK: -
+
+/// NOTE: CN docstring
+public class AuthenticationInterceptor<AuthenticatorType, AuthCredential>: RequestInterceptor where
+    AuthenticatorType: Authenticator,
+    AuthenticatorType.AuthCredential == AuthCredential
+{
+    // MARK: Helper Types
+
+    /// NOTE: CN docstring
+    public struct RefreshWindow {
+        /// NOTE: CN docstring
+        public let interval: TimeInterval
+
+        /// NOTE: CN docstring
+        public let maximumAttempts: Int
+
+        /// NOTE: CN docstring
+        public init(interval: TimeInterval = 30.0, maximumAttempts: Int = 5) {
+            self.interval = interval
+            self.maximumAttempts = maximumAttempts
+        }
+    }
+
+    private struct AdaptOperation {
+        let urlRequest: URLRequest
+        let session: Alamofire.Session
+        let completion: (Result<URLRequest, Error>) -> Void
+    }
+
+    private enum AdaptResult {
+        case adapt(AuthCredential)
+        case doNotAdapt(AuthenticationError)
+        case adaptDeferred
+    }
+
+    private struct MutableState {
+        var credential: AuthCredential?
+
+        var isRefreshing: Bool = false
```

Comment:
You can remove the explicit `Bool`.

```suggestion
        var isRefreshing = false
```

Reply:
Addressed in d7b4d1d1.